### DISCUSSION TITLE
Using Salesforce as Data source

### DISCUSSION THREAD

**Discussion by steve970** [code4recovery/12-step-meeting-list#1532] [Q&A] [answered]
Currently, I am storing all of my meeting information in Salesforce. I am in the process of using a SF report exported to a csv file that I will use to import into TSML. Has anyone had any success fully integrating Salesforce data into the TSML plugin? Any insights or documentation on how to align the different column headings would be insightful. Thanks!

**Comment by joshreisner** [comment]
i'm not aware of anyone doing this - but my suggestion would be to write some kind of script that uses the salesforce api to fetch the data and then saves it in meeting guide json format

for example, as part of the code4recovery/sheets project we wrote an importer for aa san francisco / marin that fetches data out of their airtable account and displays it as JSON.

potentially, since you're using TSML UI, you could even uninstall TSML and just have TSML UI point to that JSON feed. perhaps you would structure that importer to run every 15 minutes and save the data as a static JSON file on S3 or something…

**Reply by joshreisner** [reply]
here's more information about the format: https://github.com/code4recovery/spec

**Comment by steve970** [comment]
That sounds like something we could possibly work towards in the future. Right now, we are taking a Salesforce Report, exporting it to a CSV and importing it into the TSML. If I am in the process of importing a custom spreadsheet. Outside of this code, is there anything specific I should be doing to ensure columns align correctly?

```
if (!function_exists('tsml_import_reformat')) {
    function tsml_import_reformat($meetings) {
        // In my case, changing column headings
        // Iterate through each meeting
         foreach ($meetings as &$meeting) {
             // Check if 'Meeting Name' exists in the current meeting array
             if (isset($meeting['Meeting Name'])) {
                 // Rename 'Meeting Name' to 'Name'
                 $meeting['Name'] = $meeting['Meeting Name'];
                 // Remove the old 'Meeting Name' key
                 unset($meeting['Meeting Name']);
            }
        }
        return $meetings;
    }
}
```

**Reply by joshreisner** [reply, marked as answer]
this looks good! no need to remove the old array keys, that i'm aware of. you might need to lowercase the keys, i'm not sure. eg `$meeting['name']` 

we have a spec of all the fields that we support: https://github.com/code4recovery/spec

there is likely some trial-and-error involved here, good luck!

**Reply by steve970** [reply]
Thank you!
